plutus-ir/test/ParserSpec.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FlexibleContexts #-}
+module ParserSpec (parsing) where
+
+import           PlutusPrelude
+
+import           Common
+import           TestLib
+
+import           Control.Monad
+import           Control.Monad.Except
+import           Control.Monad.State
+
+import qualified Data.ByteString.Lazy         as BSL
+
+import           Language.PlutusIR
+import qualified Language.PlutusIR.Parser     as Parser
+
+import qualified Language.PlutusCore          as PLC
+import qualified Language.PlutusCore.Constant as PLC
+import qualified Language.PlutusCore.Pretty   as PLC
+
+import           Hedgehog                     hiding (Var)
+import qualified Hedgehog.Gen                 as Gen
+import qualified Hedgehog.Range               as Range
+
+import           Text.Megaparsec              hiding (failure)
+
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+
+propParser :: Property
+propParser = property $ do
+    prog <- forAll genProgram
+    let pretty = prettyString prog
+    ast <- case Parser.parse "test" Parser.program pretty of
+               Left e -> do
+                   annotate pretty
+                   annotate $ parseErrorPretty e
+                   failure
+               Right p -> pure p
+    pretty === prettyString ast
```

Comment:
See `tripping` in `Hedgehog` which does a bunch of this for you, including the annotation.

Reply:
The problem is, tripping is super annoying because only the second round-trip function can return a monad; in our case `parse` can't be the second function, because then you're comparing a randomly generated AST (without location info) to a  parsed AST (with location info) so it has to be the first, but then it can't live in a failure monad... lmk if you have any idea how to do this elegantly

Reply:
Well, I do something hacky [here](https://github.com/input-output-hk/plutus/blob/master/language-plutus-core/test/Spec.hs#L107), which you could borrow if you want. Otherwise I suggest at least using some of the things like `evalEither`.